### PR TITLE
Faster SD Core ML on iPhone, iPad, and Mac

### PR DESCRIPTION
**Pull request by pcuenca**

### CODE REVIEW

**Review by julien-c — COMMENTED** [review]
looks very cool!

File: fast-diffusers-coreml.md [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+authors:
+- user: pcuenq
+---
+
+# Faster Stable Diffusion with Core ML on iPhone, iPad, and Mac
+
+<!-- {blog_metadata} -->
+<!-- {authors} -->
+
+WWDC’23 (Apple Worldwide Developers Conference) was held last week. A lot of the news focused on the Vision Pro announcement during the keynote, but there’s much more to it. Like every year, WWDC week is packed with more than 200 technical sessions that dive deep inside the upcoming features across Apple operating systems and frameworks. This year we are particularly excited about changes in Core ML devoted to compression and optimization techniques. These changes make running models such as Stable Diffusion faster and with less memory use! As a taste, consider the following results I measured on my phone and Mac and comparing against the [initial port of Stable Diffusion to Core ML we discussed back in December](https://huggingface.co/blog/diffusers-coreml):
```
Comment (Vaibhavs10) [inline fast-diffusers-coreml.md:13]:
```suggestion
WWDC’23 (Apple Worldwide Developers Conference) was held last week. A lot of the news focused on the Vision Pro announcement during the keynote, but there’s much more to it. Like every year, WWDC week is packed with more than 200 technical sessions that dive deep inside the upcoming features across Apple operating systems and frameworks. This year we are particularly excited about changes in Core ML devoted to compression and optimization techniques. These changes make running [models](https://huggingface.co/apple) such as Stable Diffusion faster and with less memory use! As a taste, consider the following results I measured on my phone and Mac and comparing against the [initial port of Stable Diffusion to Core ML we discussed back in December](https://huggingface.co/blog/diffusers-coreml):
```
I think it'd be a good idea for us to brand the Apple org on HF a bit more, essentially nudging users to start using it more, also helps build the case as Apple for OS.

File: fast-diffusers-coreml.md [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+[TODO: table with numbers]
```
Comment (Vaibhavs10) [inline fast-diffusers-coreml.md:15]: I was wondering if here instead of numbers we put a pretty image comparing the diffusers (app) load time between the two iterations. It conveys the same meaning but makes it visually appealing. We can also quote that it is xX times faster.

File: fast-diffusers-coreml.md [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+<img style="border:none;" alt="Illustration of 2-bit palettization. Image credit: Apple WWDC’23 Session 'Use Core ML Tools for machine learning model compression'" src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/fast-diffusers-coreml/palettization_illustration.png" />
+
+The compressed 6-bit _weights_ cannot be used for computation, because they are just indices into a table and no longer represent the magnitude of the original weights. Therefore, Core ML needs to uncompress the palletized weights before use. In previous versions of Core ML, uncompression took place when the model was first loaded from disk, so the amount of memory used was equal to the uncompressed model size. With the new improvements, weights are kept as 6-bit numbers and converted on the fly as inference progresses from layer to layer. This might seem slow – an inference run requires a lot of uncompressing operations –, but it’s typically more efficient than preparing all the weights in 16-bit mode! The reason is that memory transfers are in the critical path of execution, and transferring less memory is faster than transferring uncompressed data.
```
Comment (Vaibhavs10) [inline fast-diffusers-coreml.md:37]: This is a brilliant explanation. Thanks for writing it in such a friendly fashion.

File: fast-diffusers-coreml.md (hunk 1 above)
Comment (Vaibhavs10) [reply fast-diffusers-coreml.md:13]: It might also be a good idea to plug this blog here instead. https://huggingface.co/blog/fast-mac-diffusers

File: fast-diffusers-coreml.md [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+* Quantization is supported using `--quantize-nbits` during conversion. You can quantize to 8, 6, 4, or even 2 bits! For best results, we recommend using 6-bit quantization, as the precision loss is small while achieving fast inference and significant memory savings. If you want to go lower than that, please check [this section](#using-less-than-6-bits) for advanced techniques.
+* Additional optimizations of the attention layers that achieve even better performance on the Neural Engine! The trick is to split the query sequences into chunks of 512 to avoid the creation of large intermediate tensors. This method is called `SPLIT_EINSUM_V2` in the code and can improve performance between 10% to 30%.
+
+In order to make it easy for everyone to take advantage of these improvements, we have converted the four official Stable Diffusion models and pushed them to the Hub. These are all the variants:
```
Comment (Vaibhavs10) [inline fast-diffusers-coreml.md:46]:
```suggestion
In order to make it easy for everyone to take advantage of these improvements, we have converted the four official Stable Diffusion models and pushed them to the [Hub](https://huggingface.co/apple). These are all the variants:
```
Just to reinforce the org.

File: fast-diffusers-coreml.md [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+If you want to apply quantization, you need the latest versions of `coremltools`, `apple/ml-stable-diffusion` and Xcode in order to do the conversion.
+* Download `coremltools` 7.0 beta from [the releases page in GitHub](https://github.com/apple/coremltools/releases).
+* Download Xcode 15.0 beta from [Apple developer site](https://developer.apple.com/).
+* Download `apple/ml-stable-diffusion` [from the repo](https://github.com/apple/ml-stable-diffusion) and follow the installation instructions.
+
+</Tip>
+
+1. Select the model you want to convert. You can train your own or choose one from the [Hugging Face Diffusers Models Gallery](https://huggingface.co/spaces/huggingface-projects/diffusers-gallery). For example, let’s convert [`prompthero/openjourney-v4`](https://huggingface.co/prompthero/openjourney-v4).
+2. Install `apple/ml-stable-diffusion` and run a first conversion using the `ORIGINAL` attention implementation like this:
```
Comment (Vaibhavs10) [inline fast-diffusers-coreml.md:88]: Would be really cool to update our space or create a variant of it. I understand the technical limitations but it might be a good idea IMO. This isn't a blocker tho, we can always do more comms when the public beta is open too.

File: fast-diffusers-coreml.md [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+    --check-output-correctness \
+    -o models/split_einsum_v2/openjourney-6-bit
+```
+
+4. Test the converted models on the desired hardware. As a rule of thumb, the `ORIGINAL` version usually works better on macOS, whereas `SPLIT_EINSUM_V2` is usually faster on iOS. For more details and additional data points, see [these tests contributed by the community](https://github.com/huggingface/swift-coreml-diffusers/issues/31) on the previous version of Stable Diffusion for Core ML.
+
+5. To integrate the desired model in your own app:
+    * If you are going to distribute the model inside the app, use the `.mlpackage` files. Note that this will increase the size of your app binary.
+    * Otherwise, you can use the compiled `Resources` to download them dynamically when your app starts.
+
```
Comment (Vaibhavs10) [inline fast-diffusers-coreml.md:135]: I think it might be a good idea to create another issue asking the community to help us benchmark on their hardware. WDYT?
I can again help create it if you want.

File: fast-diffusers-coreml.md [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+</Tip>
+
+## Using Less than 6 bits
+
+6-bit quantization is a sweet spot between model quality, model size and convenience – you just need to provide a conversion option in order to be able to quantize any pre-trained model. This is an example of _post-training compression_.
+
+The beta version of `coremltools` released last week also includes _training-time_ compression methods. The idea here is that you can fine-tune a pre-trained Stable Diffusion model and perform the weights compression while fine-tuning is taking place. This allows you to use 4- or even 2-bit compression while minimizing the loss in quality. The reason this works is because weight clustering is performed using a differentiable algorithm, and therefore we can apply the usual training optimizers to find the quantization table while minimizing model loss.
+
+We have plans to evaluate this method soon, and can’t wait to see how 4-bit optimized models work and how fast they run. If you beat us to it, please drop us a note and we’ll be happy to check 🙂
```
Comment (Vaibhavs10) [inline fast-diffusers-coreml.md:148]: Over here in the end, it'd be good to put some takeaways, IMO it always helps people reinforce the ideas. This could also be just a line or two like.. "To summarise we learned how to make Stable Diffusion faster by xX times using quantisation methods ...."

**Review by Vaibhavs10 — COMMENTED** [review]
I love how this blog post covers a lot of depth whilst making it a very easy read. Kudos @pcuenca, I left a few comments.. nothing major.

**Review by osanseviero — APPROVED** [review]
LGTM pending the TODOs and feedback on initial draft 🔥 🚀

File: fast-diffusers-coreml.md (hunk 3 above)
Comment (pcuenca) [reply fast-diffusers-coreml.md:37]: It's the most revised paragraph of the post :) Omar rightfully didn't like the first version and Apple didn't complain but I think they felt it was too hand-wavy, so it went through a couple of revisions and simplifications :) Glad is clear now!

File: fast-diffusers-coreml.md (hunk 6 above)
Comment (pcuenca) [reply fast-diffusers-coreml.md:135]: That'd be awesome!

File: fast-diffusers-coreml.md (hunk 2 above)
Comment (pcuenca) [reply fast-diffusers-coreml.md:15]: That's a very good idea, let's see if I can make it work.

**Review by Vaibhavs10 — APPROVED** [review]
LGTM @pcuenca 🚀
Let's shippp this ✨
